### PR TITLE
Refactor walker

### PR DESCRIPTION
**Pull request by ajafff**
#### PR checklist

- [x] Addresses an existing issue: #1900
- [x] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### What changes did you make?

See tracking issue #1900 

#### Is there anything you'd like reviewers to focus on?

I made a separate commit for every change. If you want I can submit them as individual PRs.

There will be a followup PR with the breaking changes for cleanup. That includes removing the is*Boundary _methods_ from their classes and only use utility function instead.

### CODE REVIEW

File: src/language/walker/walker.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
```
Comment (nchen63) [inline src/language/walker/walker.ts]: 2016

File: src/language/walker/ruleWalker.ts [hunk 2]
```diff
@@ -68,21 +64,14 @@ export class RuleWalker extends SyntaxWalker {
         }
     }
 
-    public skip(node: ts.Node) {
```
Comment (nchen63) [inline src/language/walker/ruleWalker.ts]: I'm really hope nobody is using this. It looks like it implemented with a bunch of other stuff and never used

File: src/language/walker/walker.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+
+import {RuleFailure} from "../rule/rule";
+
+export interface IWalker {
```
Comment (adidahiya) [inline src/language/walker/walker.ts]: can we put this in `walker/index.ts` instead? the import from `/.../walker/walker` doesn't look as nice as `/.../walker`

File: src/language/walker/ruleWalker.ts (hunk 2 above)
Comment (ajafff) [inline src/language/walker/ruleWalker.ts]: I readded the method.
I will remove this in the followup PR with breaking changes

**Comment by nchen63** [comment]
@ajafff thanks!

**Pull request merged**